Drawing performance issues

PetzeltA opened the pull request:
OK - this is my first try to increase performance.
The idea behind this is to detach the columns from the dom before drawing them.
They are detached when opening, closing, showing, hiding or redrawing a node, and reattached when the operation is done.
Especially the close of a node is much faster now because only open child-nodes are considered while close.
I've updated the large data test (to create 50K nodes).
In IE the speed increase is tremendous. Chrome is quite fast, so you don not feel much increase in speed. ;-)
I've also done some loop optimization, but that is not much speed increase.

deitch commented:
Ugh. Your editor reformatted almost all of the spaces/tabs. I am not at all a stickler for it, but this makes the diff nearly impossible to read.

PetzeltA commented:
I'll try to undo the formatting tomorrow. Hope I won't lose to many changes.

PetzeltA commented:
OK - the code should be more readable ;-)

deitch commented:
> OK - the code should be more readable ;-)

Thanks.  BTW, if the formatting is off, it always is a good idea to fix it. It just needs to be its own PR, so that it is possible to inspect changes.

PetzeltA commented:
Welcome. :-)
I've tested the updated tree in a clients project with 6 columns and lots of data.
I've got 2 interesting errors.

1. The first node should open, when the tree is loaded, but it stays closed.
2. Hidden node-values in the last column stay visible (the class jstree-grid-hidden is there but there is a css problem)

If I use hide and show instead of a css class, the code works.
But problem 1. is nore important

deitch commented:
This is one serious chunk of reworking. Still taking notes on it. I will comment here soon...

deitch commented:
A lot of inline comments coming now...

deitch commented on jstreegrid.js:
```diff
@@ -41,11 +41,13 @@
 			if (ids === null || ids === undefined || ids.length === 0) {
-				ret = "";
-			} else if (typeof(ids) === "string") {
+        return scope;
```
Yes, I like that it returns the original passed scope or `$()` instead of `""`

PetzeltA replied:
Returning an empty string was an error, but this is just a fallback scenario, cause ids is always provided

deitch commented on jstreegrid.js:
```diff
@@ -41,11 +41,13 @@
     },
-		findDataCell = function (uniq, ids, col) {
-			var ret = $(), columns = [].concat(col), cellId;
+    findDataCell = function (uniq, ids, col, scope) {
```
What is the point of `scope` here? In normal language, what does `findDataCell()` do differently? Before, its purpose was: given: a tree (`uniq`); 0, 1 or more `id` of nodes,; 1 or more columns `col`; then find the actual cells that contain that data. Basically, it is like finding a series of cells on an Excel spreadsheet given the rows (`id`) and columns (`col`). 

What does it do now with `scope`?

PetzeltA replied:
You need the scope, because we want wo search outside the document wihtin the detached columns! Just using $() - as it was before - would not find the nodes in the detached columns.
So when we search outside the document, we have to provide a scope to search in!

deitch replied:
OK, got it. I thought that was it, but wasn't sure. So `scope` limits the scope in which to search. If not provided, just use the document's DOM. Makes sense.

deitch commented:
Please let me finish all of them before you comment. I will indicate here when done.

deitch commented on jstreegrid.js:
```diff
@@ -216,7 +218,7 @@
 				}
 				// find which column our tree shuld go in
-				for (i=0;i<s.columns.length;i++) {
+        for (var i = 0, len = s.columns.length;i<len;i++) {
```
You _really_ like setting a `var len` to the length in loops, don't you? I don't care, but I don't think it buys all that much. Leave it for now, I guess.

PetzeltA replied:
Hey - it was about profiling. So caching length is faster. ;-)
But yes - there is not much of a speed gain.

deitch replied:
I expect almost nothing here. A good engine should optimize this away (although not sure, since it would need to evaluate each time... whatever, leave it).

deitch commented on jstreegrid.js:
```diff
@@ -247,7 +249,8 @@
 						'.jstree-grid-header-themeroller {border: 0; padding: 1px 3px;}',
-						'.jstree-grid-header-regular {position:relative; background-color: #EBF3FD; z-index: 1;}',
+            '.jstree-grid-header-regular {position:relative; background-color: #EBF3FD; z-index: 1;}',
+            '.jstree-grid-hidden {display: none;}',
```
Yep, classic `.hidden` CSS pattern

deitch commented on jstreegrid.js:
```diff
@@ -456,11 +459,25 @@
-        findDataCell(this.uniq, data.node.id, this._gridSettings.gridcols).remove();
-				// hide any children
-        this._hide_grid(data.node);
+      .on("hide_node.jstree", $.proxy(function (e, data) {
```
Should this be `on("hide_node.jstree")`? In this loop, you are **removing** css class `"jstree-grid-hidden"`. Is this supposed to be `.on("show_node.jstree")`?

PetzeltA replied:
Yes - you are right! I'll update the code

deitch commented on jstreegrid.js:
```diff
@@ -812,16 +829,76 @@
+      }
+      return s4() + s4() + '-' + s4() + '-' + s4() + '-' +
+        s4() + '-' + s4() + s4() + s4();
+    }
```
OK, a guid generator. Not sure I know why we use it, but ok.

deitch commented on jstreegrid.js:
```diff
@@ -812,16 +829,76 @@
+		 * Override open_node to detach the columns before redrawing child-nodes, and do reattach them afterwarts
+		 */
+    this.open_node = function (obj, callback, animation) {
+      var id = $.isArray(obj) ? _guid() : this.get_node(obj).id;
```
Why are we generating a guid here?

PetzeltA replied:
Following idea: When we get an array of nodes, then we generate a guid, cause we want to detach the columns once  and we want not to generate a pseudo guid by combining the ids of the nodes in the array.
Then all nodes are to be opened, and at least we have to reattach the nodes.

deitch replied:
Is the purpose of the id (either the id of a single node or the generated guid) just to give a pointer when detaching and reattaching nodes?

PetzeltA replied:
yes - because if you do a deep redraw and the root + some child-nodes are opened, redraw_node is called for the root and then for every visible child node, but you have to detach/reattach the columns just once. so the id of the first node wins.

deitch replied:
> you have to detach/reattach the columns just once

How does having `id` make that happen with a deep redraw?

PetzeltA replied:
when you detach the columns, you detach them with the id.
a reattach can only happen with the same id. so in a deep redraw child nodes cannot accidently reattach the columns

deitch replied:
So if I call `reattach()` with the wrong ID, it just ignores it? I see that [here](https://github.com/deitch/jstree-grid/pull/232/files#diff-aef40b6b2c133ba8c110573ca3043de3R866)

Where does the deep redraw occur?

PetzeltA replied:
Yes - wrong id will be ignored.
The user can call the redraw-method of the jstree component. You need to do it - especially when you hide multiple nodes.
I use the tree + grid components to show a task hierarchy. completed tasks are hidden on load, but the user can show them. So if the user clicks the show/hide-button, the object-state the completed tasks is switched an a complete redraw is initialized.

deitch replied:
Got it.

deitch commented on jstreegrid.js:
```diff
@@ -812,16 +829,76 @@
 		 */
-		this.redraw_node = function(obj, deep, is_callback, force_render) {
+    this.redraw_node = function (obj, deep, is_callback, force_render) {
+      var id = $.isArray(obj) ? _guid() : this.get_node(obj).id;
```
Same question. Why are we generating a guid here?

PetzeltA replied:
Same as in open_node, but here I need to detach the columns before calling parents redraw_node!

deitch commented on jstreegrid.js:
```diff
@@ -812,16 +829,76 @@
 			// next prepare the grid for a redrawn node - but only if ths node is not hidden (search does that)
-			if (obj && !this.is_hidden(obj)) {
-				this._prepare_grid(obj);
+      if (obj) {
```
So this is the core of the rendering improvements? Detach, redo when not visual, reattach?

PetzeltA replied:
Yes - always do DOM manipulation on detached nodes

deitch commented on jstreegrid.js:
```diff
@@ -812,16 +829,76 @@
 			// first allow the parent to redraw the node
 			obj = parent.redraw_node.call(this, obj, deep, is_callback, force_render);
 			// next prepare the grid for a redrawn node - but only if ths node is not hidden (search does that)
-			if (obj && !this.is_hidden(obj)) {
```
Why do we no longer check if it is hidden? Is it because we redraw every one, but some are kept hidden, just like the tree node?

PetzeltA replied:
yes - we do a redraw on every node. just keep them hidden.

deitch commented on jstreegrid.js:
```diff
@@ -857,16 +934,30 @@
+      this._detachColumns(node.id);
+      var children = node.children ? node.children : [], cols = this._gridSettings.columns || [], treecol = this._gridSettings.treecol;
+      // try to remove all children
+      for (var i = 0, len = children.length; i < len; i++) {
```
Is this deep nested loop necessary? How does this new `_hide_grid()` differ from the original, which just did `findDataCell(this.uniq,children,this._gridSettings.gridcols).remove();`?

PetzeltA replied:
The problem of the old _hide_grid was that it used children_d - so every childnode (opened or not) was included in the selector. So doing a recursion and just regarding open nodes is much more efficient (see - the updated test (Large Data Set Search).

deitch replied:
More efficient? I would have expected it to be less.

deitch commented on jstreegrid.js:
```diff
@@ -857,16 +934,30 @@
 		this.holdingCells = {};
-		this.popHoldingCells = function (obj,col,hc) {
+    this.popHoldingCells = function (obj, col, hc) {
+      if (obj.state.hidden) { return $(); }
```
If the object is hidden, then doesn't that mean that the item will remain on the holding cells stack? Should we not be popping it but not showing it?

PetzeltA replied:
Maybe you're right. I haven't got the holding-cells concept by now.
The drawing  procedures in the original control, don't have such a concept. Perhaps it would be better to handle the whole drwaing like in the tree and use nested unsigned lists instead of divs. The idea of nested elements is a very clean. especially when removing/hiding child elements!

deitch replied:
Holding cells are just a mechanism to hold cells that need to be drawn, but cannot be drawn because their parents are not yet drawn.

Two points:

1. With the detach/attach, holding cells become less important. You basically can render all of them and then sort them. But the algorithm does get messy.
2. The grid used to be (several years ago) `div` and `span` elements appended to the end of the tree, so sorting, opening/closing, etc. was all handled as part of the tree. The downside was that getting horizontal positioning correct - let alone column rearrangement and resizing - was nearly impossible. So we switched to a true grid. Trade-offs.

deitch commented on jstreegrid.js:
```diff
@@ -1032,22 +1123,28 @@
-				last = a;
+        last = a;
+
+        // calculate position ids once
```
I like these, pulling them out. It makes a ton of sense.

deitch commented on jstreegrid.js:
```diff
@@ -1032,22 +1123,28 @@
 					}
 					col = cols[i];
-					dataCell = mw.children("div:eq("+i+")");
+          column = $(this._domManipulation.columns[i]); //Geht the detached column not mw.children("div:eq("+i+")");
```
Not sure I understand this. Previously, we basically got the `dataCell`, which was the column `div`. Is this the same thing, just using it from the detached as opposed to the attached `mw`?

PetzeltA replied:
i renamed dataCell to column, cause mw.children are columns!
And then we take the detached column.

deitch replied:
So `column === dataCell` conceptually? Just 

1. Use the detached column
2. Give it a new name

If that is it, then sure, makes sense.

deitch commented on jstreegrid.js:
```diff
@@ -1032,22 +1123,28 @@
-					// add a column class to the dataCell
-					dataCell.addClass(ccl);
+					// add a column class to the Column
+          column.addClass(ccl);
```
Similar: if `column` is just the equivalent of `dataCell` but using the detached, then I understand this.

PetzeltA replied:
you got it! :-)

deitch commented on jstreegrid.js:
```diff
@@ -1090,7 +1187,7 @@
 					}
 
-					last = findDataCell(uniq, lid, i);
+          last = findDataCell(uniq, lid, i, column);
```
Comes back to the `scope` question. What does adding `column` to `findDataCell()` do? Is it basically to allow it to search in detached? In other words, without `scope`, it would search in the rendered DOM, so now it searches in a detached node set?

PetzeltA replied:
yep - that's it!

deitch commented on jstreegrid.js:
```diff
@@ -1111,7 +1208,15 @@
+
+          // we need to check the hidden-state and see if we need to hide the node
+          if (objData.state.hidden) {
+            last.addClass("jstree-grid-hidden");
```
Nice.

deitch commented on jstreegrid.js:
```diff
@@ -1169,7 +1271,7 @@
 							last.insertBefore(gridCellNext);
 						} else {
-							last.appendTo(dataCell);
+							last.appendTo(column);
```
Same:  if `column` is just the equivalent of `dataCell` but using the detached, then I understand this.

deitch commented:
Detach/reattach: Is there any concern about flashing and loss of animation?

* Flashing: could the detach and then reattach lead to jerky/flashing?
* animation: The tree node cells are animated, does this get in the way?

deitch commented:
> Hidden node-values in the last column stay visible (the class jstree-grid-hidden is there but there is a css problem)

What do Chrome debugger or Firebug show? I always found CSS easiest to debug there.

deitch commented:
> The first node should open, when the tree is loaded, but it stays closed

Is the tree node open and the grid nodes missing? Or all closed?

deitch commented:
OK, all me inline comments and questions done. Whew.

PetzeltA commented:
hehe - ok. do you want me to comment via the reply statements or to do it in a one lengthly reply?

deitch commented:
Whatever makes it easier.

PetzeltA commented:
I didn't see any flashing. And animation never really worked.

deitch commented:
I am waiting to see if you are done commenting before trying to read through them all... :-)

PetzeltA commented:
I'm done commenting and I fixed some problems, though the redraw_node problem was not a real one, cause redraw_node is always called with a node an not an array - so the code was ok ther!

deitch commented:
> I didn't see any flashing. And animation never really worked.

For jstree? Sure, it works all the time. Look at the samples on the right of https://www.jstree.com

Unless you mean animation never worked for grid, in which case you are correct.

PetzeltA commented:
just meant the animation of the grid.
as we do not detach the original grid-column (no need to do that, cause the grid replaints effective), there are no animation issues/flickering in the main column

deitch commented:
So let's summarize current state:

1. The holding cells issue raised above needs to be fixed
2. The error with the `on("hide_node")` needs to be fixed (unless you got it already)
3. Anything else?

Once those are done, let's run a full suite of tests. I assume you have, but would like to also.

deitch commented:
I think that last one will work, but full tests will tell.

That `hide_node`/`show_node` done?

PetzeltA commented:
Ok - I just fixed the hoding cells issue, though we should find out if the holding-cells things are really needed now. If not, this would speed the grid a lot!.
The second on("hide_node") is fixed and a show is done.
I think that should be all. Let's do some testing. I do them in a real project. ;-)

deitch commented:
> though we should find out if the holding-cells things are really needed now.

Agreed. But let's get this working, merged in, and then remove extraneous bits. Separate merged = reduced risk of error.

> Let's do some testing

Yup.

>  I do them in a real project. ;-)

Ha!

deitch commented:
I just ran all of the tests (one of these days I must automate all of this). I found lots of test issues, but most unrelated to this change. Here were the ones I found unique:

1. `api-hide-show-test`: "Hide Nodes" hides the entire column rather than individual rows, and "Add product" messes up the order.
2. `dnd-test`: drag works, and it drops, but it never completes the drop. It appears in the new tree, not the old one, and is stuck to the pointer. 
3. `search-test`: 1st tree works fine; second tree does not show grid at all and matches everything; third tree does not open or show anything, even matched.

deitch commented:
Question: Does detaching and reattaching the column do just the data, or the column header too? Probably should do just the data.

PetzeltA commented:
There is also an issue with popHoldingCells.
When there are hidden nodes and you add neu ones or redraw certain nodes it sometimes messes up the order. When I change popHoldingCells to its old logic (removing the delete statements) everything is fine.

PetzeltA commented:
Question: Does detaching and reattaching the column do just the data, or the column header too? Probably should do just the data.
Answer: The whole column (with header) is detached (because it' one div). Manipulation occurs without the header.

deitch commented:
> Answer: The whole column (with header) is detached (because it' one div). Manipulation occurs without the header.

That explains some of the columns detaching and not reappearing. We probably should do it separately. It would make debugging easier and reduce the chances of columns shrinking/reappearing.

PetzeltA commented:
The open-icon of the root also disappears, when clicking on a colum cell

deitch commented:
> The open-icon of the root also disappears, when clicking on a colum cell

With which test?

PetzeltA commented:
In my customers project - but it should be visible with every test. Just user IE11 as your browser.
BTW: How do you test everything? By running testmaster.html in the local file system or through a web server?

deitch commented:
> Just use IE11 as your browser

Would have to spin up a Windows VM.

> By running test-master.html in the local file system or through a web server?

Both. For some browsers, just file system. But some don't allow `file:///` URLs (notably Chrome), so I spin up a python or node or php Web server.

PetzeltA commented:
api-hide-show-test: "Hide Nodes" hides the entire column rather than individual rows, and "Add product" messes up the order.
This is corredted now: There was an error in hide_node/show_node, and I've restored getHoldingCells, cause the popHoldingCells with it's delete statemens did not work when there were hidden nodes.

deitch commented:
FYI GitHub lets you quote easily by putting the character ">" at the beginning of a line followed by a space and then text.

deitch commented:
> api-hide-show-test: This is correct now.

Looks good over here. So what is left? `dnd-test` and `search-test`.

deitch commented:
Is that all of the open issues?

PetzeltA commented:
> dnd-test: drag works, and it drops, but it never completes the drop. It appears in the new tree, not the old one, and is stuck to the pointer.

fixed
>search-test: 1st tree works fine; second tree does not show grid at all and matches everything; third tree does not open or show anything, even matched.

fixed
The problem was that sometimes _prepare_grid was called without a call to _detachCells. I wrapped all direct calls to _prepare_grid with calls to detach/reattach and also - if someone forgets - created a fallback in _prepare_grid.

PetzeltA commented:
>Is that all of the open issues?

I hope so :-)

deitch commented:
OK, time to rerun the tests.

deitch commented:
FYI, there are about five tests that are broken in general, but not due to these changes.

deitch commented:
This looks good. All the tests that work pass now. 

Is there still an issue with the root node that you raised earlier? Or is that one fixed now as well?

deitch commented:
Oh, add your name to the contributors in `package.json` please and bump the patch number?

deitch commented on package.json:
```diff
@@ -1,43 +1,47 @@
 	"name": "jstreegrid",
 	"description": "grid plugin for jstree",
-	"version": "3.8.2",
+	"version": "3.9.1",
```
Why 3.9.1? Why not 3.8.3? According to semver, this has no API change so should remain same minor version.

deitch commented:
Hell of a job, @PetzeltA !

deitch commented:
`npm publish` done. Wow.

PetzeltA commented:
Thx - but I still am not satisfied with the outcome. If I have some time, I'll change the columns to nested ul/li items. this should speed up the code even more. :-)

deitch commented:
New PR, new branch.

I get what you are saying. It does feel like it would be a duplicate of the tree itself, though. I would like to find a real way to square that corner.
